### ISSUE TITLE
Do not attempt to renew certificate no longer used

### ISSUE THREAD

**Issue by Yajo** [containous/traefik#3376]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I booted one service in one server, behind Traefik, configured with docker backend and correctly labeled.

Everything worked fine:

- Traefik contacted LE
- Traefik obtained cert
- Traefik renewed cert after awhile, when needed
- Traefik :guitar: :heart: 

But, after some time, i moved the service to another equally-configured server.

### What did you expect to see?

Traefik in old server stop renewing cert if no container is using it actively.

### What did you see instead?

These logs:

```
INFO[2018-05-23T16:30:05Z] Renewing certificate from LE : {Main:example.com SANs:[]} 
legolog: 2018/05/23 16:30:05 [INFO][odoo.turbointernacional.com] acme: Trying renewal with -1447 hours remaining
legolog: 2018/05/23 16:30:05 [INFO][odoo.turbointernacional.com] acme: Obtaining bundled SAN certificate
legolog: 2018/05/23 16:30:05 [INFO][odoo.turbointernacional.com] AuthURL: https://acme-v01.api.letsencrypt.org/acme/authz/******
legolog: 2018/05/23 16:30:05 [INFO][odoo.turbointernacional.com] acme: Could not find solver for: dns-01
legolog: 2018/05/23 16:30:05 [INFO][odoo.turbointernacional.com] acme: Trying to solve HTTP-01
ERRO[2018-05-23T16:31:35Z] Error renewing certificate from LE: acme: Error 403 - urn:acme:error:unauthorized - Invalid response from http://example.com/.well-known/acme-challenge/****** [83.48.28.165]: 404
Error Detail:
        Validation for example.com:80
        Resolved to:
                *.*.*.*
        Used: *.*.*.*
```

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.5.3
Codename:     cancoillotte
Go version:   go1.9.4
Built:        2018-02-27_02:47:04PM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

Docker backend, CLI flags configuration... but I don't think is relevant for this issue, where everything is working as expected.

**Comment by minhdanh** [comment]
I'm seeing this behavior on Traefik 1.7-rc2 as well

**Comment by kumy** [comment]
Is there any known workaround for this?

**Comment by MaxWinterstein** [comment]
Any news on this? I also ran into an issue where my traefik had requested a huge number of certs for non existing frontends.

Bash oneliner guy i am, i hacked some commands together to clean those old certs out of the acme.json. I use the api endpoint of the dashboard for this. I leave this here for anyone, but please think before you do something. I only use kubernetes backend.

## **How I removed unused certs**

**_I take no warranty for your copy paste job! What worked for me, maybe fail on your environment_**

I simply jump directly into my traefik container and did the folling:

* Install dependencies
```bash
apk update && apk add jq curl
```

* Fetch existing frontends
```bash
curl -s "https://<USERNAME>:<PASSWORD>@<TRAEFIK_DASHBOARD_URL>/api" | jq ".kubernetes.frontends" | jq "keys" | jq -r ".[]" | sed "s/\/.*//" | uniq > existing_frontends
```

* Fetch existing certs
```bash
cat /acme/acme.json | jq ".Certificates" | jq ".[]" | jq ".Domain"  | jq -r ".Main" | sort | uniq > existing_certs
```

* now get the certs that need to be removed
```bash
diff existing_certs existing_frontends | tail -n +4 | grep "^-" | sed "s/^-//" > certs_to_remove
```

* and let jq remove them
```bash 
cp /acme/acme.json /acme/acme.json.new
cat certs_to_remove | xargs -i sh -c "jq 'del(.Certificates[]| select(.Domain.Main == \"{}\"))' /acme/acme.json.new > /acme/acme.json.new2; mv /acme/acme.json.new2 /acme/acme.json.new"
```

* verify the new file
```bash
cat /acme/acme.json.new | jq ".Certificates[].Domain.Main"
```

* change permission
```bash
chmod 600 /acme/acme.json.new
```

* backup and overwrite
```bash
cp /acme/acme.json /acme/acme.json.bak
echo "remove next # - think before you do something"
#cp /acme/acme.json.new /acme/acme.json
```
**_I take no warranty for your copy paste job!_**

**Comment by puco** [comment]
This is especially annoying, when the certificates are stored in KV store (consul in our case) which limits the size of the acme.json object. We spin up instances on demand and tear them down after couple of days. But the certificates stay in the file and eventually preventing new certificates from being created.

The only workaround for me here is to stop traefik, semi-manually remove the obsolete certificates, push the new file to the KV store and start traefik again.

For a non-existing URL it does not make sense to renew and can be removed. Once the URL would be used again traefik can request the certificate again.

**Comment by strajansebastian** [comment]
also confirm this. 
it would be nice to have an API call that can automatically delete the certs from the store (file/consul/other)

this are the logs from the traefik:v1.7.9 docker container:

```
{"log":"time=\"2019-04-07T17:50:16Z\" level=info msg=\"Renewing certificate from LE : {Main:My.Removed.Domain.TLD SANs:[]}\"\n","stream":"stdout","time":"2019-04-07T17:50:16.346915604Z"}
{"log":"time=\"2019-04-07T17:50:16Z\" level=info msg=\"legolog: [INFO] [My.Removed.Domain.TLD] acme: Trying renewal with 688 hours remaining\"\n","stream":"stdout","time":"2019-04-07T17:50:16.347232814Z"}
{"log":"time=\"2019-04-07T17:50:16Z\" level=info msg=\"legolog: [INFO] [My.Removed.Domain.TLD] acme: Obtaining bundled SAN certificate\"\n","stream":"stdout","time":"2019-04-07T17:50:16.34740912Z"}
{"log":"time=\"2019-04-07T17:50:16Z\" level=info msg=\"legolog: [INFO] [My.Removed.Domain.TLD] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz/AUTHZ_TOKEN_NOT_SHARED\"\n","stream":"stdout","time":"2019-04-07T17:50:16.659316163Z"}
{"log":"time=\"2019-04-07T17:50:16Z\" level=info msg=\"legolog: [INFO] [My.Removed.Domain.TLD] acme: Could not find solver for: tls-alpn-01\"\n","stream":"stdout","time":"2019-04-07T17:50:16.659442867Z"}
{"log":"time=\"2019-04-07T17:50:16Z\" level=info msg=\"legolog: [INFO] [My.Removed.Domain.TLD] acme: use http-01 solver\"\n","stream":"stdout","time":"2019-04-07T17:50:16.659449967Z"}
{"log":"time=\"2019-04-07T17:50:16Z\" level=info msg=\"legolog: [INFO] [My.Removed.Domain.TLD] acme: Trying to solve HTTP-01\"\n","stream":"stdout","time":"2019-04-07T17:50:16.659453967Z"}
{"log":"time=\"2019-04-07T17:50:21Z\" level=error msg=\"Error renewing certificate from LE: acme: Error -\u003e One or more domains had a problem:\\n[My.Removed.Domain.TLD] acme: error: 403 :: urn:ietf:params:acme:error:unauthorized :: Invalid response from http://My.Removed.Domain.TLD/.well-known/acme-challenge/ACME_CHALLANGE_TOKEN_NOT_SHARED [1.3.3.7]: \\\"\u003chtml\u003e\\\\r\\\\n\u003chead\u003e\u003ctitle\u003e403 Forbidden\u003c/title\u003e\u003c/head\u003e\\\\r\\\\n\u003cbody bgcolor=\\\\\\\"white\\\\\\\"\u003e\\\\r\\\\n\u003ccenter\u003e\u003ch1\u003e403 Forbidden\u003c/h1\u003e\u003c/center\u003e\\\\r\\\\n\u003chr\u003e\u003ccenter\u003e\\\", url: \\n\"\n","stream":"stdout","time":"2019-04-07T17:50:21.45825789Z"}
```

This is a script wrote a while back in which I made manual changes in order to removed the old/bad/unused/migrated certs from the consul acme.json:
Removed manually all reference for the targeted domains.

After this a manual push of the cert is required made. Ensure that you don't corrupt your acme.json while editing (you'll still have the original backups at FILE_ORIGINAL_BASE64)

You need to run this on one of your consul servers (script tested on ubuntu 16.04) (in some cases had to reboot the entire setup to make the changes visible, still didn't found a stable/easy/logical way of making the update).
Take the script from below as a guideline and for the first run I would recommend a step by step run so you'll understand what happens.
Also look after variables that start with CHANGE_ME and make the appropriate changes.

```
#!/bin/bash

# 
# IF YOU RUN THIS SCRIPT ENSURE THAT NO CERTIFICATE UPDATES ARE MADE DURING THIS PERIOD
# YOU RISK TO DELETE/OVERRIDE NEWLY GENERATED CERTIFICATES
#

BASE_CERT_PATH=traefik
CERT_PATH=${BASE_CERT_PATH}/acme/account/object

BASE_FILE_NAME=/root/consul_dump.`date +%Y-%m-%d_%H-%M-%S`
FILE_ORIGINAL_BASE64=${BASE_FILE_NAME}.1.base64
FILE_ORIGINAL_JSON=${BASE_FILE_NAME}.2.json
FILE_MODIFIED_JSON=${BASE_FILE_NAME}.3.json.modified
FILE_MODIFIED_BASE64=${BASE_FILE_NAME}.4.base64.modified

# get data out of consul - in base64 gzip enconding

# use this for HTTTPS authentication
CONSUL_AUTH_PARAMS="-ca-path=/consul/tls/ca.pem -client-cert=/consul/tls/consul.pem -client-key=/consul/tls/consul-key.pem -http-addr=https://127.0.0.1:8443 -tls-server-name=CHANGE_ME_CONSUL-HA-HOSTNAME.TLD"
echo `docker exec CHANGE_ME_CONSUL-DOCKER-CONTAINER-NAME consul kv get $CONSUL_AUTH_PARAMS -base64 $CERT_PATH` > $FILE_ORIGINAL_BASE64
cat $FILE_ORIGINAL_BASE64 | base64 --decode | gzip -dc | jq . > $FILE_ORIGINAL_JSON
cp $FILE_ORIGINAL_JSON $FILE_MODIFIED_JSON

# make manual changes
vim $FILE_MODIFIED_JSON

# convert back to base64 gzip and store for push upstream
cat $FILE_MODIFIED_JSON | gzip -c | base64 -w 0 > $FILE_MODIFIED_BASE64 

# put data back to consul store
echo 'THIS IS DONE'
echo 'MANUALLY RUN: '

echo "cat $FILE_MODIFIED_BASE64 | /var/lib/docker/overlay2/CHANGE_ME_CONSUL-LAYER-THAT-CONTAINS-THE-CONSUL-BINARY/diff/bin/consul kv put $CONSUL_AUTH_PARAMS -base64 $CERT_PATH -"
```

**Comment by nitrique** [comment]
+1 same here

**Comment by Jensderond** [comment]
For what its worth I created a Makefile that is tested with Traefik 2.2.x
This is mostly based on the [comment](https://github.com/containous/traefik/issues/3376#issuecomment-462795574) of @MaxWinterstein 

```make
acmefile = acme.json
traefik_dashboard = <TRAEFIK_DASHBOARD_URL>
auth_user = <USERNAME>
auth_password = <PASSWORD>

.SILENT: clean
.PHONY: clean
clean:
    curl -s "https://$(auth_user):$(auth_password)@$(traefik_dashboard)/api/http/routers" | jq -r ".[]" | jq ".rule" | sed "s/\"Host(\`//g;s/\`)\"//g" | uniq > existing_frontends;
    cat $(acmefile) | jq ".default.Certificates[].domain.main" | sort | uniq | sed "s/\"//g" > existing_certs;
    awk 'NR==FNR{a[$$0];next}!($$0 in a)' existing_frontends existing_certs > certs_to_remove;
    cp $(acmefile) $(acmefile).new;
    cat certs_to_remove | xargs -I'{}' -i sh -c "jq 'del(.default.Certificates[]| select(.domain.main == \"{}\"))' $(acmefile).new > $(acmefile).new2; mv $(acmefile).new2 $(acmefile).new";
    chmod 600 $(acmefile).new;
    chown traefik:docker $(acmefile).new;
    mv $(acmefile) $(acmefile).bak;
    mv $(acmefile).new $(acmefile);
    rm certs_to_remove existing_certs existing_frontends;
    docker-compose restart;
```

**Comment by derjohn** [comment]
Wow,
that issue is still open after 2 years? With Traefik you can trigger the Let's Encrypt rate limit pretty fast, just have 5 certs in acme.json that can't be renewed, because the domain moved the DNS wo somewhere else I believe a non-existent domain does not trigger it (NXDOMAIN), but domains pointing to another http endpoint, will trigger the rate-limit.
Removing from ACME Json will cause a traefik start, or at least kill -HUP, so causing downtime.
Is this the "upsell reason" for buying traefik EE ?

**Comment by derjohn** [comment]
Just a side-note: There was a related ticket, where there is an API call mentioned that should be implemented: https://github.com/traefik/traefik/issues/7082

**Comment by MarkErik** [comment]
I am also worried about hitting the rate limit for for URLs that may have been pointed to other addresses.

It has been ~5 months since the last comment, have any of you found a better solution? (or is it to just run one of the scripts?)

**Comment by derjohn** [comment]
@MarkErik Basically it's still one of those scripts. I ran traefik in k8s and started migrating to ingress-nginx and cert manger: That scales better (multi-pod with Certs) and has better cert-management.

**Comment by Aaron-Ritter** [comment]
> Just a side-note: There was a related ticket, where there is an API call mentioned that should be implemented: #7082

it was just a suggestion but there is no concrete issue related to it, did you find anything related to that?

**Comment by 205g0** [comment]
@ldez is anyone on this issue?

> I ran traefik in k8s and started migrating to ingress-nginx and cert manger: That scales better (multi-pod with Certs) and has better cert-management.

@derjohn better cert-management? would you mind to elaborate? and how was the general setup of both nginx and cert-manager compared to Traefik 2.0?

**Comment by AndrewSav** [comment]
> @derjohn better cert-management? would you mind to elaborate?
@205g0 well you just need to read up on / try the cert-manager. Certs are not buried in a json file that can be in any of the supported PVs somewhere, and instead are exposed as first class Kubernetes Secret objects,  there are traceable Order / Challenge objects for acme (with lifecycle events) and crs/cert for all certs. Also there is support for Vault, self-signed and external CAs that's absent in traefik. All that amounts to better cert-management.

**Comment by 205g0** [comment]
@AndrewSav thanks! Yesterday, I had checked cert-manager but somehow I was afraid or just too lazy to set it up. Which is the best ingress to pair it with? The Kubernetes nginx or the one from Nginx Inc. or an entirely different ingress?

**Comment by AndrewSav** [comment]
@205g0 cert-manager works well with all of them. I personally use traefik as ingress in my kuberentes clusters, it's working well for me.

**Comment by MaxWinterstein** [comment]
> @ldez is anyone on this issue?
> 
> > I ran traefik in k8s and started migrating to ingress-nginx and cert manger: That scales better (multi-pod with Certs) and has better cert-management.
> 
> @derjohn better cert-management? would you mind to elaborate? and how was the general setup of both nginx and cert-manager compared to Traefik 2.x?
> 
> Edit: Just found that e.g. https://voyagermesh.com/docs/v2021.04.24-rc.0/guides/certificate/delete/ has a cert delete feature, this would be also handy for Traefik.

Can definitely recommend the [cert-manager](https://github.com/jetstack/cert-manager/) and [acme-dns](https://github.com/joohoi/acme-dns)  way. Using this to get wildcart lets-encrypt certs refreshed works nice for me.

ditched traefik for [ambassador](https://github.com/datawire/ambassador) a while ago and never looked back.

**Comment by 205g0** [comment]
@MaxWinterstein cert-manager is def the way to go, even it is paired with Traefik, better than Traefik's built-in resolver. Ambassador is on my list, I checked Contour and Gloo before because people compared them all the time to Ambassador and yeah. I got Contour running, nice docs but it seems not be very feature-rich but somewhat a community. Gloo look best on paper but I couldn't get it to run, so yeah... you're happy with Ambassador? Any drawbacks?

**Comment by ZVilusinsky** [comment]
Aaaand the apparent fix PR was just killed :-/

**Comment by emilevauge** [comment]
> Aaaand the apparent fix PR was just killed :-/

It wasn't killed at all:
- the maintainer team did a review of the PR and made [some comments, asked questions](https://github.com/traefik/traefik/pull/10782#issuecomment-2270819629) which is the normal procedure.
- None of this those comments has been addressed by the author of the PR.
- The author removed the permission to reuse his code to take over his PR.
- After 2 months of inactivity, we made the decision to close it as there was no sign of life and no way to take over.

Reviewing and discussing PRs is a normal process of any OSS project. This is one of the guarantee to keep the code clean, maintainable, safe, etc 
A contributor should be ready to chat about his work I guess :)

To summarize, although we do agree the actual ACME implementation could be improved, we also think this can raise some questions that should be discussed :)

**Comment by ZVilusinsky** [comment]
From my reading , all relevant questions were answered https://github.com/traefik/traefik/pull/10782#issuecomment-2161207749 . 

While the PR might not be the best solution, it is A solution for a more than six year old (and 2 major versions) pain point for your users. And while the pain is less than it used to be with consul provider (size limitation), it is still there.

**Comment by emilevauge** [comment]
Some early questions were answered indeed, but the [additional review made by the team](https://github.com/traefik/traefik/pull/10782#issuecomment-2270819629) wasn't. I'm also sorry for this situation, but I don't see how we could proceed differently if the author is unresponsive.

**Comment by ieugen** [comment]
Wow, this is a really old bug.

**Comment by Alestrix** [comment]
I had this bite me, too!

I moved my domains to another server and the whole installation with it, including `acme.json` and only using docker labels for configuration of the domains. The main domains (like `service.mydomain.tld`) stayed the same, but the SANs switched from `service.oldserver.mydomain.tld` to `service.newserver.mydomain.tld`. Traefik still tried to renew the certificates for the `oldserver` FQDNs, preventing the whole cert update process from succeeding and thus causing the certs for even the main domains to expire.
Deleting acme.json and making traefik re-request all certs did the trick, but I was holding my breath, hoping not to run into letsencrypt's throttling due to the amount of certs.

**Comment by colinmollenhour** [comment]
FYI, wrote a script that removes expired certs: https://gist.github.com/colinmollenhour/6db6014aa7a72406d32f1e8e782d29e6
Note, this does not remove unused certs that are not expired - so I agree that should still be implemented, just thought this might be helpful to some.

**Comment by kale1d0code** [comment]
Our hosting has this problem a lot.
are there any docker services out in the wild that can manage the acme.json file?

**Comment by almereyda** [comment]
Like [ldez/traefik-certs-cleaner](https://github.com/ldez/traefik-certs-cleaner)?

**Comment by kale1d0code** [comment]
> Like [ldez/traefik-certs-cleaner](https://github.com/ldez/traefik-certs-cleaner)?

Not quite, however better than nothing thanks for sharing, I'll check it out :)

 I was wondering if there was something a bit more automated.
maybe listening on the docker socket, collect all the containers with the traefik.enable label, then determining the domain names and cross referencing the acme.json file

**Comment by ffried** [comment]
>I was wondering if there was something a bit more automated.
>maybe listening on the docker socket, collect all the containers with the traefik.enable label, then determining the domain names and cross referencing the acme.json file

This was exactly what I was also missing. Didn't have enough time to write this myself from scratch, but Copilot helped me bit here (and I've refined some bits later): [sersoft-gmbh/traefik-acme-storage-cleaner](https://github.com/sersoft-gmbh/traefik-acme-storage-cleaner)
It uses the official Traefik repository as source for the data types (to make sure the JSON file has the correct structure).
It simply removes all expired certificates - so not a full "check what's still used" but merely a "prevent renewal errors" solution.
There is a downside, however: Traefik does *not* watch that file and keeps its contents in memory. So to cleanup the file, traefik needs to be stopped. At least I didn't find a better way. I'm currently running this container nightly (stopping Traefik beforehand).

**Comment by ShadowJonathan** [comment]
What exactly is this issue being blocked on atm? Attention/triage from maintainers?
